internal/ls/utilities.go

```diff
@@ -711,6 +1006,512 @@ func nodeEndsWith(n *ast.Node, expectedLastToken ast.Kind, sourceFile *ast.Sourc
+		if parentType := parent.Type(); parentType != nil && parentType.Kind == ast.KindTypeReference {
+			return parentType.AsTypeReferenceNode().TypeName
+		}
+	}
+	// /**/readonly [|name|][]
+	if node.Kind == ast.KindReadonlyKeyword && parent.Kind == ast.KindTypeOperator && parent.AsTypeOperatorNode().Operator == ast.KindReadonlyKeyword {
+		if parentType := parent.Type(); parentType != nil && parentType.Kind == ast.KindArrayType && parentType.AsArrayTypeNode().ElementType.Kind == ast.KindTypeReference {
+			return parentType.AsArrayTypeNode().ElementType.AsTypeReferenceNode().TypeName
+		}
+	}
+
+	if !forRename {
+		// /**/new [|name|]
+		// /**/void [|name|]
+		// /**/void obj.[|name|]
+		// /**/typeof [|name|]
+		// /**/typeof obj.[|name|]
+		// /**/await [|name|]
+		// /**/await obj.[|name|]
+		// /**/yield [|name|]
+		// /**/yield obj.[|name|]
+		// /**/delete obj.[|name|]
+		if node.Kind == ast.KindNewKeyword && parent.Kind == ast.KindNewExpression ||
+			node.Kind == ast.KindVoidKeyword && parent.Kind == ast.KindVoidExpression ||
+			node.Kind == ast.KindTypeOfKeyword && parent.Kind == ast.KindTypeOfExpression ||
+			node.Kind == ast.KindAwaitKeyword && parent.Kind == ast.KindAwaitExpression ||
+			node.Kind == ast.KindYieldKeyword && parent.Kind == ast.KindYieldExpression ||
+			node.Kind == ast.KindDeleteKeyword && parent.Kind == ast.KindDeleteExpression {
+			if expr := parent.Expression(); expr != nil {
+				return ast.SkipOuterExpressions(expr, ast.OEKAll)
+			}
+		}
+
+		// left /**/in [|name|]
+		// left /**/instanceof [|name|]
+		if (node.Kind == ast.KindInKeyword || node.Kind == ast.KindInstanceOfKeyword) && parent.Kind == ast.KindBinaryExpression && parent.AsBinaryExpression().OperatorToken == node {
+			return ast.SkipOuterExpressions(parent.AsBinaryExpression().Right, ast.OEKAll)
+		}
+
+		// left /**/as [|name|]
+		if node.Kind == ast.KindAsKeyword && parent.Kind == ast.KindAsExpression {
+			if asExprType := parent.Type(); asExprType != nil && asExprType.Kind == ast.KindTypeReference {
+				return asExprType.AsTypeReferenceNode().TypeName
+			}
+		}
+
+		// for (... /**/in [|name|])
+		// for (... /**/of [|name|])
+		if node.Kind == ast.KindInKeyword && parent.Kind == ast.KindForInStatement ||
+			node.Kind == ast.KindOfKeyword && parent.Kind == ast.KindForOfStatement {
+			return ast.SkipOuterExpressions(parent.AsForInOrOfStatement().Expression, ast.OEKAll)
+		}
+	}
+
+	return node
+}
+
+func getAdjustedLocationForDeclaration(node *ast.Node, forRename bool, sourceFile *ast.SourceFile) *ast.Node {
+	if node.Name() != nil {
+		return node.Name()
+	}
+	if forRename {
+		return nil
+	}
+	switch node.Kind {
+	case ast.KindClassDeclaration, ast.KindFunctionDeclaration:
+		// for class and function declarations, use the `default` modifier
+		// when the declaration is unnamed.
+		if node.Modifiers() != nil {
```

Comment:
This is currently only called one place, inside a check that the original node is a modifier, and the node in this function is its parent. So this check isn't strictly needed. (But is probably a port, and probably a good idea in case this function is used elsewhere.)